src/services/review/RepoCheckerService.ts

```diff
@@ -0,0 +1,234 @@
+          if (!fileContent) {
+            return
+          }
+          const html = await marked.parse(fileContent)
+          const dom = new JSDOM(html)
+          const anchorTags = dom.window.document.querySelectorAll("a")
+          forEach(anchorTags, (tag) => {
+            const href = tag.getAttribute("href")
+            const text = tag.textContent
+            if (href) {
+              console.log(href)
+              const encodedHref = this.documentPathDecoder(href)
+              if (
+                !this.isExternalLinkOrPageRef(encodedHref) &&
+                !setOfAllMediaAndPagesPath.has(encodedHref) &&
+                // Amplify supports adding of trailing slash
+                !setOfAllMediaAndPagesPath.has(`${encodedHref}/`)
+              ) {
+                console.log(setOfAllMediaAndPagesPath)
+                this.persistLogs(repo, [
+                  new SiteCheckerError(
+                    `Broken link ${href} that links ${text} found in ${fileName} which has permalink ${filePermalink}`
+                  ),
+                ])
+              }
+            }
+          })
+          const imgTags = dom.window.document.querySelectorAll("img")
+          forEach(imgTags, (tag) => {
+            const src = tag.getAttribute("src")
+            if (src) {
+              const encodedSrc = this.documentPathDecoder(src)
+              if (
+                !this.isExternalLinkOrPageRef(encodedSrc) &&
+                !setOfAllMediaAndPagesPath.has(encodedSrc)
+              ) {
+                this.persistLogs(repo, [
+                  new SiteCheckerError(
+                    `Broken image ${src} found in ${fileName} which has permalink ${filePermalink}`
+                  ),
+                ])
+              }
+            }
+          })
+        })
+        console.log("Site checker done")
+        return ok(undefined)
+      })
+    try {
+      console.log(await checker)
+    } catch (error) {
+      console.log(error)
+    }
+  }
+
+  isExternalLinkOrPageRef(link: string) {
+    return (
+      link.startsWith("http") ||
+      link.startsWith("www") ||
+      link.startsWith("mailto:") ||
+      link.startsWith("#")
+    )
+  }
+
+  documentPathDecoder(documentPath: string) {
+    for (let i = 0; i < ALLOWED_FILE_EXTENSIONS.length; i += 1) {
+      if (documentPath.endsWith(ALLOWED_FILE_EXTENSIONS[i])) {
+        return decodeURIComponent(documentPath)
+      }
+    }
+    return documentPath
+  }
+
+  getListOfMarkdownFiles(
+    repoPath: string
+  ): ResultAsync<Set<string>, SiteCheckerError> {
+    const setOfAllMarkdownFiles = new Set<string>()
+    return ResultAsync.fromPromise(
+      fs.promises.readdir(repoPath, { recursive: true }),
+      (error) => new SiteCheckerError(`${error}`)
+    ).andThen((files) => {
+      files
+        .filter((file) => !file.startsWith(".") && file.endsWith(".md"))
+        .map((file) => {
+          setOfAllMarkdownFiles.add(file)
```

Comment:
This is abit confusing - can we just combine the `map` and `andThen` and scope the set locally?

Reply:
ok if this is done upstream ah?

Reply:
sure